src/lib/keys/KeyManager.ts

```diff
@@ -459,16 +495,31 @@ class KeyManager {
     })
   }
 
+  /* ============ HELPERS =============== */
   getKey(name: string): Buffer {
     return this.derivedKeys[name]
   }
 
-  isLoaded(): boolean {
+  isLoaded(name: string): boolean {
     if (this.derivedKeys[name]) {
       return true
     }
     return false
   }
+
+  private writeMetadata(): void {
```

Comment:
I think the `fs` here is being hardcoded, and in preparation for nativescript we should actually abstract the `fs` into a dependency injected property of these Polykey classes.

That's step 1.

The other issue is resource management.

Resource management of external resources is similar to memory management. If we use RAII, we tie resources to memory management and traditional OOP garbage collection.

If not, then we should use the analogue of "manual memory management" but for resources. And in JS that means `try` plus `finally` clauses to open and close descriptors.

In fact in other languages there will be generally a higher level `file` object, and then you can use that instead. This exists in Python and C actually, with the lower level fd object existing internally.

Reply:
The below functions are side-effectful convenience functions, but they don't compose really well. Firstly they are synchronous thus blocking the entire thread when reading/writing files. But also they assume alot, the assume the existence of `fs` that matches that API that is available in the outside static context, and then they assume that the file they are writing to can be opened/closed and nothing else cares about that file afterwards. What if you needed to create that file ahead of time with specific permissions first?

In the name of composition, you should think about what you can do abstract some of this.

Of course in the name of practice, you eventually need to close the loop, tie the knot so to speak. Usually that's the IO layer in haskell. And so that's when all the loose ends must have closure, and you plug in the real world.

Finally doing this makes it easier to test/mock.

Reply:
All of these methods now use the file system that is passed into the class in the same way that `EncryptedFS` takes a file system parameters. This should cover us for nativescript.

Reply:
Great awesome.